src/linear_algebra/annihilating_polynomial.lean

```diff
@@ -0,0 +1,207 @@
+  rw aeval_def,
+  apply h,
+end
+
+end semiring
+
+section field
+
+variables {𝕜 A : Type*} [field 𝕜] [ring A] [algebra 𝕜 A]
+variable (𝕜)
+
+open submodule
+
+/-- Since `𝕜[X]` is a principal ideal domain there is a polynomial `g` such that
+ `span 𝕜 {g} = ann_ideal a`. This picks some generator.
+ We prefer the monic generator of the ideal -/
+noncomputable def ann_ideal_generator (a : A) : 𝕜[X] :=
+let g := is_principal.generator $ ann_ideal 𝕜 a
+  in (C g.leading_coeff⁻¹) * g
+
+/-- ann_ideal_generator 𝕜 a is indeed a generator -/
+lemma span_singleton_ann_ideal_generator (a : A) :
+  span 𝕜[X] {ann_ideal_generator 𝕜 a} = ann_ideal 𝕜 a :=
+begin
+  by_cases (is_principal.generator $ ann_ideal 𝕜 a) = 0,
+  { rw ← is_principal.eq_bot_iff_generator_eq_zero at h,
+   simp only [ann_ideal_generator, h], simp, rw ← is_principal.eq_bot_iff_generator_eq_zero, },
+  { simp only [ann_ideal_generator, ann_ideal, alg_hom.to_ring_hom_eq_coe, ideal.submodule_span_eq],
+    rw ideal.span_singleton_mul_left_unit,
+    { exact is_principal.span_singleton_generator _ },
+    { rw [is_unit_C, is_unit_iff_ne_zero],
+      apply inv_ne_zero,
+      rw [ne.def, leading_coeff_eq_zero_iff_deg_eq_bot, degree_eq_bot],
+      rw [ann_ideal, alg_hom.to_ring_hom_eq_coe] at h,
+      apply h, }, },
+end
+
+/-- The annihilating ideal generator is a member of the annihilating ideal,
+  following submodule.generator_mem -/
+lemma ann_ideal_generator_mem (a : A) : ann_ideal_generator 𝕜 a ∈ ann_ideal 𝕜 a :=
+let I := ann_ideal 𝕜 a,
+    g := submodule.is_principal.generator I in
+  I.mul_mem_left (C g.leading_coeff⁻¹) (submodule.is_principal.generator_mem I)
+
+/-- sourced from submodule.mem_iff_eq_smul_generator -/
+lemma mem_iff_eq_smul_ann_ideal_generator {p : 𝕜[X]} (a : A) :
+  p ∈ ann_ideal 𝕜 a ↔ ∃ s : 𝕜[X], p = s • ann_ideal_generator 𝕜 a :=
+by simp_rw [@eq_comm _ p, ← mem_span_singleton, span_singleton_ann_ideal_generator]
+
+/-- sourced from submodule.eq_bot_iff_generator_eq_zero -/
+lemma eq_bot_iff_ann_ideal_generator_eq_zero (a : A) :
+  ann_ideal 𝕜 a = (⊥ : ideal 𝕜[X]) ↔ ann_ideal_generator 𝕜 a = (0 : 𝕜[X]) :=
+begin
+  rw ← span_singleton_ann_ideal_generator,
+  apply @span_singleton_eq_bot 𝕜[X] 𝕜[X] _ _ _ (ann_ideal_generator 𝕜 a),
+end
+
+/-- The obvious scaling of a non-zero polynomial gives a monic polynomial.
+ This can be moved to another file, since it does not use anything about ann_ideal. -/
+lemma mul_monic_of_ne_zero (g : 𝕜[X]) (hg : g ≠ 0) : monic ((C g.leading_coeff⁻¹) * g) :=
+begin
+  unfold monic,
+  rw leading_coeff_mul (C g.leading_coeff⁻¹) g,
+  rw leading_coeff_C g.leading_coeff⁻¹,
+  rw mul_comm,
+  rw mul_inv_cancel,
+  apply leading_coeff_ne_zero.2 hg,
+end
+
+/-- The generator we chose for the annihilating ideal is monic when the ideal is non-zero -/
+lemma monic_of_ann_ideal_generator (a : A) (hg : (ann_ideal_generator 𝕜 a : 𝕜[X]) ≠ 0) :
+  monic (ann_ideal_generator 𝕜 a : 𝕜[X]) :=
+begin
+  unfold ann_ideal_generator,
+  simp,
```

Comment:
Please squeeze the nonterminal `simp`.

Reply:
@riccardobrasca I do not know how to do this. There is a let expression. I tried to work with it, but my rewrites downstream in the proof don't go into the let expression. I don't know how else to inline that expression so I can continue